lib/delayed/backend/sqs.rb

```diff
@@ -85,20 +113,45 @@ def save
           if @attributes[:handler].blank?
             raise "Handler missing!"
           end
-          payload = JSON.dump(@attributes)
+          payload = ::DelayedJobSqs::Document.sqs_safe_json_dump(@attributes)
 
-          # Resend the message before deleting from queue to ensure if resend fails, message stays and job can be retried when visibility timeout is exceeded
-          sqs.queues.named(queue_name).send_message(payload, :delay_seconds  => @delay)
-          @msg.delete if @msg # TODO:  potential problem here in that there may be multiple copies of this message on the q since SQS guarantees to write at least once
-          
+          @msg.delete if @msg
+
+          maxed_delay = [900, @delay + 5 + attempts ** 4].min
+
+          if buffering?
+            send_later({ message_body: payload, delay_seconds: maxed_delay })
+          else
+            sqs.queues.named(queue_name).send_message(payload, :delay_seconds  => @delay )
+          end
           true
         end
 
         def save!
           save
         end
 
-        # Destroy the job; that is, delete it from the SQS queue and don't save it anywhere.
+        def send_later(message)
+          buffer[@queue_name] = [[]] unless buffer[@queue_name]
+          current_buffer = buffer[@queue_name]
+
+          current_buffer_size = current_buffer.last.reduce(0) { |m, msg| m + msg[:message_body].bytesize }
+          if current_buffer_size + message[:message_body].bytesize >= ::DelayedJobSqs::Document::MAX_SQS_MESSAGE_SIZE_IN_BYTES ||
```

Comment:
I feel like it would still be too long, but I would happily do things to make this smaller.